File: examples/go/dynamic_programming.go
package main

import "fmt"

/*
Dynamic Programming Pattern Examples

Dynamic Programming solves complex problems by breaking them down into simpler subproblems.
It stores the results of subproblems to avoid redundant calculations.

Common approaches:
1. Top-down (Memoization): Recursion + Cache
2. Bottom-up (Tabulation): Iterative + Table

Time Complexity: Usually O(n) to O(n²)
Space Complexity: O(n) to O(n²) for memoization/tables
*/

// Fibonacci calculates nth Fibonacci number using bottom-up DP
// Time: O(n), Space: O(1)
func Fibonacci(n int) int {
	if n <= 1 {
		return n
	}

	prev1, prev2 := 0, 1

	for i := 2; i <= n; i++ {
		current := prev1 + prev2
		prev1, prev2 = prev2, current
	}

	return prev2
}

// ClimbStairs counts ways to climb n stairs (1 or 2 steps at a time)
// Time: O(n), Space: O(1)
func ClimbStairs(n int) int {
	if n <= 2 {
		return n
	}

	prev1, prev2 := 1, 2

	for i := 3; i <= n; i++ {
		current := prev1 + prev2
		prev1, prev2 = prev2, current
	}

	return prev2
}

// HouseRobber robs houses to maximize money without robbing adjacent houses
// Time: O(n), Space: O(1)
func HouseRobber(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	if len(nums) == 1 {
		return nums[0]
	}

	prev1, prev2 := 0, nums[0]

	for i := 1; i < len(nums); i++ {
		current := max(prev2, prev1+nums[i])
		prev1, prev2 = prev2, current
	}

	return prev2
}

// CoinChange finds minimum coins needed to make amount
// Time: O(amount * len(coins)), Space: O(amount)
func CoinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = amount + 1 // Initialize with impossible value
	}

	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if coin <= i {
				dp[i] = min(dp[i], dp[i-coin]+1)
			}
		}
	}

	if dp[amount] > amount {
		return -1
	}
	return dp[amount]
}

// LongestIncreasingSubsequence finds length of longest increasing subsequence
// Time: O(n²), Space: O(n)
func LongestIncreasingSubsequence(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	n := len(nums)
	dp := make([]int, n)
	for i := range dp {
		dp[i] = 1
	}

	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			if nums[j] < nums[i] {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
	}

	result := dp[0]
	for _, val := range dp {
		if val > result {
			result = val
		}
	}

	return result
}

// Knapsack01 solves 0/1 Knapsack problem
// Time: O(n * capacity), Space: O(capacity)
func Knapsack01(weights, values []int, capacity int) int {
	n := len(weights)
	dp := make([]int, capacity+1)

	for i := 0; i < n; i++ {
		// Traverse backwards to avoid using updated values
		for w := capacity; w >= weights[i]; w-- {
			dp[w] = max(dp[w], dp[w-weights[i]]+values[i])
		}
	}

	return dp[capacity]
}

// LongestCommonSubsequence finds length of longest common subsequence
// Time: O(m * n), Space: O(min(m, n))
func LongestCommonSubsequence(text1, text2 string) int {
	m, n := len(text1), len(text2)

	// Use space-optimized version
	prev := make([]int, n+1)

	for i := 1; i <= m; i++ {
		curr := make([]int, n+1)
		for j := 1; j <= n; j++ {
			if text1[i-1] == text2[j-1] {
				curr[j] = prev[j-1] + 1
			} else {
				curr[j] = max(prev[j], curr[j-1])
			}
		}
		prev = curr
	}

	return prev[n]
}

// EditDistance finds minimum edit distance (insert, delete, replace)
// Time: O(m * n), Space: O(min(m, n))
func EditDistance(word1, word2 string) int {
	m, n := len(word1), len(word2)

	// Space-optimized version
	prev := make([]int, n+1)
	for j := 0; j <= n; j++ {
		prev[j] = j
	}

	for i := 1; i <= m; i++ {
		curr := make([]int, n+1)
		curr[0] = i
		for j := 1; j <= n; j++ {
			if word1[i-1] == word2[j-1] {
				curr[j] = prev[j-1]
			} else {
				curr[j] = 1 + min(prev[j], min(curr[j-1], prev[j-1]))
			}
		}
		prev = curr
	}

	return prev[n]
}

// WordBreak checks if string can be segmented using dictionary words
// Time: O(n² * m), Space: O(n)
func WordBreak(s string, wordDict []string) bool {
	wordSet := make(map[string]bool)
	for _, word := range wordDict {
		wordSet[word] = true
	}

	dp := make([]bool, len(s)+1)
	dp[0] = true

	for i := 1; i <= len(s); i++ {
		for j := 0; j < i; j++ {
			if dp[j] && wordSet[s[j:i]] {
				dp[i] = true
				break
			}
		}
	}

	return dp[len(s)]
}

// MaxProductSubarray finds maximum product of contiguous subarray
// Time: O(n), Space: O(1)
func MaxProductSubarray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	maxProd, minProd, result := nums[0], nums[0], nums[0]

	for i := 1; i < len(nums); i++ {
		num := nums[i]

		// Handle negative numbers by swapping max and min
		if num < 0 {
			maxProd, minProd = minProd, maxProd
		}

		maxProd = max(num, maxProd*num)
		minProd = min(num, minProd*num)

		result = max(result, maxProd)
	}

	return result
}

// UniquePathsWithObstacles counts unique paths in grid with obstacles
// Time: O(m * n), Space: O(n)
func UniquePathsWithObstacles(obstacleGrid [][]int) int {
	if len(obstacleGrid) == 0 || len(obstacleGrid[0]) == 0 || obstacleGrid[0][0] == 1 {
		return 0
	}

	m, n := len(obstacleGrid), len(obstacleGrid[0])
	dp := make([]int, n)
	dp[0] = 1

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if obstacleGrid[i][j] == 1 {
				dp[j] = 0
			} else if j > 0 {
				dp[j] += dp[j-1]
			}
		}
	}

	return dp[n-1]
}

// Helper functions
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// Test functions
func testFibonacci() {
	fmt.Println("Testing Fibonacci calculation...")

	testCases := []struct {
		input    int
		expected int
	}{
		{0, 0},
		{1, 1},
		{10, 55},
		{15, 610},
	}

	for _, tc := range testCases {
		result := Fibonacci(tc.input)
		if result != tc.expected {
			panic(fmt.Sprintf("Fibonacci test failed for %d: expected %d, got %d",
				tc.input, tc.expected, result))
		}
	}

	fmt.Println("✅ Fibonacci tests passed")
}

func testClimbStairs() {
	fmt.Println("Testing climb stairs...")

	testCases := []struct {
		input    int
		expected int
	}{
		{1, 1},
		{2, 2},
		{3, 3},
		{4, 5},
	}

	for _, tc := range testCases {
		result := ClimbStairs(tc.input)
		if result != tc.expected {
			panic(fmt.Sprintf("Climb stairs test failed for %d: expected %d, got %d",
				tc.input, tc.expected, result))
		}
	}

	fmt.Println("✅ Climb stairs tests passed")
}

func testHouseRobber() {
	fmt.Println("Testing house robber...")

	testCases := []struct {
		input    []int
		expected int
	}{
		{[]int{1, 2, 3, 1}, 4},
		{[]int{2, 7, 9, 3, 1}, 12},
		{[]int{5}, 5},
	}

	for _, tc := range testCases {
		result := HouseRobber(tc.input)
		if result != tc.expected {
			panic(fmt.Sprintf("House robber test failed for %v: expected %d, got %d",
				tc.input, tc.expected, result))
		}
	}

	fmt.Println("✅ House robber tests passed")
}

func testCoinChange() {
	fmt.Println("Testing coin change...")

	testCases := []struct {
		coins    []int
		amount   int
		expected int
	}{
		{[]int{1, 3, 4}, 6, 2},
		{[]int{2}, 3, -1},
		{[]int{1}, 0, 0},
	}

	for _, tc := range testCases {
		result := CoinChange(tc.coins, tc.amount)
		if result != tc.expected {
			panic(fmt.Sprintf("Coin change test failed for coins=%v, amount=%d: expected %d, got %d",
				tc.coins, tc.amount, tc.expected, result))
		}
	}

	fmt.Println("✅ Coin change tests passed")
}

func testLIS() {
	fmt.Println("Testing longest increasing subsequence...")

	testCases := []struct {
		input    []int
		expected int
	}{
		{[]int{10, 9, 2, 5, 3, 7, 101, 18}, 4},
		{[]int{0, 1, 0, 3, 2, 3}, 4},
		{[]int{7, 7, 7, 7, 7, 7, 7}, 1},
	}

	for _, tc := range testCases {
		result := LongestIncreasingSubsequence(tc.input)
		if result != tc.expected {
			panic(fmt.Sprintf("LIS test failed for %v: expected %d, got %d",
				tc.input, tc.expected, result))
		}
	}

	fmt.Println("✅ LIS tests passed")
}

func testKnapsack() {
	fmt.Println("Testing 0/1 knapsack...")

	weights := []int{1, 3, 4, 5}
	values := []int{1, 4, 5, 7}
	expected := 9

	result := Knapsack01(weights, values, 7)
	if result != expected {
		panic(fmt.Sprintf("Knapsack test failed: expected %d, got %d", expected, result))
	}

	fmt.Println("✅ Knapsack tests passed")
}

func testLCS() {
	fmt.Println("Testing longest common subsequence...")

	testCases := []struct {
		text1    string
		text2    string
		expected int
	}{
		{"abcde", "ace", 3},
		{"abc", "abc", 3},
		{"abc", "def", 0},
	}

	for _, tc := range testCases {
		result := LongestCommonSubsequence(tc.text1, tc.text2)
		if result != tc.expected {
			panic(fmt.Sprintf("LCS test failed for '%s', '%s': expected %d, got %d",
				tc.text1, tc.text2, tc.expected, result))
		}
	}

	fmt.Println("✅ LCS tests passed")
}

func testEditDistance() {
	fmt.Println("Testing edit distance...")

	testCases := []struct {
		word1    string
		word2    string
		expected int
	}{
		{"horse", "ros", 3},
		{"intention", "execution", 5},
	}

	for _, tc := range testCases {
		result := EditDistance(tc.word1, tc.word2)
		if result != tc.expected {
			panic(fmt.Sprintf("Edit distance test failed for '%s', '%s': expected %d, got %d",
				tc.word1, tc.word2, tc.expected, result))
		}
	}

	fmt.Println("✅ Edit distance tests passed")
}

func testWordBreak() {
	fmt.Println("Testing word break...")

	testCases := []struct {
		s        string
		wordDict []string
		expected bool
	}{
		{"leetcode", []string{"leet", "code"}, true},
		{"applepenapple", []string{"apple", "pen"}, true},
		{"catsandog", []string{"cats", "dog", "sand", "and", "cat"}, false},
	}

	for _, tc := range testCases {
		result := WordBreak(tc.s, tc.wordDict)
		if result != tc.expected {
			panic(fmt.Sprintf("Word break test failed for '%s': expected %v, got %v",
				tc.s, tc.expected, result))
		}
	}

	fmt.Println("✅ Word break tests passed")
}

func testMaxProduct() {
	fmt.Println("Testing max product subarray...")

	testCases := []struct {
		input    []int
		expected int
	}{
		{[]int{2, 3, -2, 4}, 6},
		{[]int{-2, 0, -1}, 0},
		{[]int{-2, 3, -4}, 24},
	}

	for _, tc := range testCases {
		result := MaxProductSubarray(tc.input)
		if result != tc.expected {
			panic(fmt.Sprintf("Max product test failed for %v: expected %d, got %d",
				tc.input, tc.expected, result))
		}
	}

	fmt.Println("✅ Max product tests passed")
}

func main() {
	fmt.Println("Testing Dynamic Programming Pattern...")
	testFibonacci()
	testClimbStairs()
	testHouseRobber()
	testCoinChange()
	testLIS()
	testKnapsack()
	testLCS()
	testEditDistance()
	testWordBreak()
	testMaxProduct()
	fmt.Println("\n🎉 All Dynamic Programming tests passed!")
}

File: examples/python/dynamic_programming.py
#!/usr/bin/env python3
"""
Dynamic Programming Pattern Examples

Dynamic Programming solves complex problems by breaking them down into simpler subproblems.
It stores the results of subproblems to avoid redundant calculations.

Common approaches:
1. Top-down (Memoization): Recursion + Cache
2. Bottom-up (Tabulation): Iterative + Table

Time Complexity: Usually O(n) to O(n²)
Space Complexity: O(n) to O(n²) for memoization/tables
"""

from typing import List, Dict
from functools import lru_cache


def fibonacci(n: int) -> int:
    """
    Calculate nth Fibonacci number using bottom-up DP.
    
    Args:
        n: The position in Fibonacci sequence
        
    Returns:
        The nth Fibonacci number
        
    Time: O(n), Space: O(1)
    """
    if n <= 1:
        return n
    
    prev1, prev2 = 0, 1
    
    for i in range(2, n + 1):
        current = prev1 + prev2
        prev1, prev2 = prev2, current
    
    return prev2


def climb_stairs(n: int) -> int:
    """
    Count ways to climb n stairs (1 or 2 steps at a time).
    
    Args:
        n: Number of stairs
        
    Returns:
        Number of ways to climb stairs
        
    Time: O(n), Space: O(1)
    """
    if n <= 2:
        return n
    
    prev1, prev2 = 1, 2
    
    for i in range(3, n + 1):
        current = prev1 + prev2
        prev1, prev2 = prev2, current
    
    return prev2


def house_robber(nums: List[int]) -> int:
    """
    Rob houses to maximize money without robbing adjacent houses.
    
    Args:
        nums: Money in each house
        
    Returns:
        Maximum money that can be robbed
        
    Time: O(n), Space: O(1)
    """
    if not nums:
        return 0
    if len(nums) == 1:
        return nums[0]
    
    prev1, prev2 = 0, nums[0]
    
    for i in range(1, len(nums)):
        current = max(prev2, prev1 + nums[i])
        prev1, prev2 = prev2, current
    
    return prev2


def coin_change(coins: List[int], amount: int) -> int:
    """
    Find minimum coins needed to make amount.
    
    Args:
        coins: Available coin denominations
        amount: Target amount
        
    Returns:
        Minimum coins needed, or -1 if impossible
        
    Time: O(amount * len(coins)), Space: O(amount)
    """
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0
    
    for i in range(1, amount + 1):
        for coin in coins:
            if coin <= i:
                dp[i] = min(dp[i], dp[i - coin] + 1)
    
    return dp[amount] if dp[amount] != float('inf') else -1


def longest_increasing_subsequence(nums: List[int]) -> int:
    """
    Find length of longest increasing subsequence.
    
    Args:
        nums: Input array
        
    Returns:
        Length of LIS
        
    Time: O(n²), Space: O(n)
    """
    if not nums:
        return 0
    
    n = len(nums)
    dp = [1] * n
    
    for i in range(1, n):
        for j in range(i):
            if nums[j] < nums[i]:
                dp[i] = max(dp[i], dp[j] + 1)
    
    return max(dp)


def knapsack_01(weights: List[int], values: List[int], capacity: int) -> int:
    """
    0/1 Knapsack problem - maximize value within weight capacity.
    
    Args:
        weights: Weight of each item
        values: Value of each item
        capacity: Maximum weight capacity
        
    Returns:
        Maximum value achievable
        
    Time: O(n * capacity), Space: O(capacity)
    """
    n = len(weights)
    dp = [0] * (capacity + 1)
    
    for i in range(n):
        # Traverse backwards to avoid using updated values
        for w in range(capacity, weights[i] - 1, -1):
            dp[w] = max(dp[w], dp[w - weights[i]] + values[i])
    
    return dp[capacity]


def longest_common_subsequence(text1: str, text2: str) -> int:
    """
    Find length of longest common subsequence.
    
    Args:
        text1: First string
        text2: Second string
        
    Returns:
        Length of LCS
        
    Time: O(m * n), Space: O(min(m, n))
    """
    m, n = len(text1), len(text2)
    
    # Use space-optimized version
    prev = [0] * (n + 1)
    
    for i in range(1, m + 1):
        curr = [0] * (n + 1)
        for j in range(1, n + 1):
            if text1[i-1] == text2[j-1]:
                curr[j] = prev[j-1] + 1
            else:
                curr[j] = max(prev[j], curr[j-1])
        prev = curr
    
    return prev[n]


def edit_distance(word1: str, word2: str) -> int:
    """
    Find minimum edit distance (insert, delete, replace).
    
    Args:
        word1: First word
        word2: Second word
        
    Returns:
        Minimum edit distance
        
    Time: O(m * n), Space: O(min(m, n))
    """
    m, n = len(word1), len(word2)
    
    # Space-optimized version
    prev = list(range(n + 1))
    
    for i in range(1, m + 1):
        curr = [i]
        for j in range(1, n + 1):
            if word1[i-1] == word2[j-1]:
                curr.append(prev[j-1])
            else:
                curr.append(1 + min(prev[j], curr[j-1], prev[j-1]))
        prev = curr
    
    return prev[n]


def word_break(s: str, word_dict: List[str]) -> bool:
    """
    Check if string can be segmented using dictionary words.
    
    Args:
        s: Input string
        word_dict: Dictionary of valid words
        
    Returns:
        True if string can be segmented
        
    Time: O(n² * m), Space: O(n)
    """
    word_set = set(word_dict)
    dp = [False] * (len(s) + 1)
    dp[0] = True
    
    for i in range(1, len(s) + 1):
        for j in range(i):
            if dp[j] and s[j:i] in word_set:
                dp[i] = True
                break
    
    return dp[len(s)]


def max_product_subarray(nums: List[int]) -> int:
    """
    Find maximum product of contiguous subarray.
    
    Args:
        nums: Input array
        
    Returns:
        Maximum product
        
    Time: O(n), Space: O(1)
    """
    if not nums:
        return 0
    
    max_prod = min_prod = result = nums[0]
    
    for i in range(1, len(nums)):
        num = nums[i]
        
        # Handle negative numbers by swapping max and min
        if num < 0:
            max_prod, min_prod = min_prod, max_prod
        
        max_prod = max(num, max_prod * num)
        min_prod = min(num, min_prod * num)
        
        result = max(result, max_prod)
    
    return result


@lru_cache(maxsize=None)
def unique_paths_with_obstacles_memo(m: int, n: int, obstacles: tuple) -> int:
    """
    Count unique paths with obstacles using memoization.
    
    Args:
        m: Grid height
        n: Grid width
        obstacles: Tuple of obstacle positions
        
    Returns:
        Number of unique paths
        
    Time: O(m * n), Space: O(m * n)
    """
    obstacle_set = set(obstacles)
    
    def dfs(row: int, col: int) -> int:
        if row >= m or col >= n or (row, col) in obstacle_set:
            return 0
        if row == m - 1 and col == n - 1:
            return 1
        
        return dfs(row + 1, col) + dfs(row, col + 1)
    
    return dfs(0, 0)


# Test functions
def test_fibonacci():
    """Test Fibonacci calculation"""
    assert fibonacci(0) == 0
    assert fibonacci(1) == 1
    assert fibonacci(10) == 55
    assert fibonacci(15) == 610
    print("✅ Fibonacci tests passed")


def test_climb_stairs():
    """Test stair climbing"""
    assert climb_stairs(1) == 1
    assert climb_stairs(2) == 2
    assert climb_stairs(3) == 3
    assert climb_stairs(4) == 5
    print("✅ Climb stairs tests passed")


def test_house_robber():
    """Test house robber"""
    assert house_robber([1, 2, 3, 1]) == 4
    assert house_robber([2, 7, 9, 3, 1]) == 12
    assert house_robber([5]) == 5
    print("✅ House robber tests passed")


def test_coin_change():
    """Test coin change"""
    assert coin_change([1, 3, 4], 6) == 2
    assert coin_change([2], 3) == -1
    assert coin_change([1], 0) == 0
    print("✅ Coin change tests passed")


def test_lis():
    """Test longest increasing subsequence"""
    assert longest_increasing_subsequence([10, 9, 2, 5, 3, 7, 101, 18]) == 4
    assert longest_increasing_subsequence([0, 1, 0, 3, 2, 3]) == 4
    assert longest_increasing_subsequence([7, 7, 7, 7, 7, 7, 7]) == 1
    print("✅ LIS tests passed")


def test_knapsack():
    """Test 0/1 knapsack"""
    weights = [1, 3, 4, 5]
    values = [1, 4, 5, 7]
    assert knapsack_01(weights, values, 7) == 9
    print("✅ Knapsack tests passed")


def test_lcs():
    """Test longest common subsequence"""
    assert longest_common_subsequence("abcde", "ace") == 3
    assert longest_common_subsequence("abc", "abc") == 3
    assert longest_common_subsequence("abc", "def") == 0
    print("✅ LCS tests passed")


def test_edit_distance():
    """Test edit distance"""
    assert edit_distance("horse", "ros") == 3
    assert edit_distance("intention", "execution") == 5
    print("✅ Edit distance tests passed")


def test_word_break():
    """Test word break"""
    assert word_break("leetcode", ["leet", "code"]) == True
    assert word_break("applepenapple", ["apple", "pen"]) == True
    assert word_break("catsandog", ["cats", "dog", "sand", "and", "cat"]) == False
    print("✅ Word break tests passed")


def test_max_product():
    """Test max product subarray"""
    assert max_product_subarray([2, 3, -2, 4]) == 6
    assert max_product_subarray([-2, 0, -1]) == 0
    assert max_product_subarray([-2, 3, -4]) == 24
    print("✅ Max product tests passed")


if __name__ == "__main__":
    print("Testing Dynamic Programming Pattern...")
    test_fibonacci()
    test_climb_stairs()
    test_house_robber()
    test_coin_change()
    test_lis()
    test_knapsack()
    test_lcs()
    test_edit_distance()
    test_word_break()
    test_max_product()
    print("\n🎉 All Dynamic Programming tests passed!")